### ISSUE TITLE
Files ignored by “make dist”

### ISSUE THREAD

**Issue by dilyanpalauzov** [tpm2-software/tpm2-tss#1348]
“make dist” ignores:
```
lib/tss2-esys.def
lib/tss2-mu.def
lib/tss2-tcti-mssim.def
lib/tss2-sys.def
lib/tss2-tcti-tbs.def
.gitattributes
.appveyor.yml
doc/TSS block diagram.png
Dockerfile
.dockerignore
bootstrap
src/tss2-sys/tss2-sys.vcxproj
src/tss2-tcti/tcti-tbs.c
src/tss2-tcti/tss2-tcti-tbs.vcxproj.filters
src/tss2-tcti/tss2-tcti-tbs.vcxproj
src/tss2-tcti/tss2-tcti-mssim.sln
src/tss2-tcti/tcti-tbs.h
src/tss2-tcti/tss2-tcti-mssim.vcxproj.filters
src/tss2-tcti/tss2-tcti-mssim.vcxproj
src/tss2-tcti/tss2-tcti-tbs.sln
src/tss2-esys/tss2-esys.vcxproj.filters
src/tss2-esys/tss2-esys.vcxproj
src/tss2-esys/tss2-esys.sln
src/tss2-mu/tss2-mu.vcxproj
MAINTAINERS
include/tss2/tss2_tcti_tbs.h
script/gen_fuzz.py
script/fuzz-log-compiler.sh
tpm2-tss.sln
.gitignore
.travis.yml
```
Consider tweaking the tests to first call “make dist”, unpack and call the tests on the unpacked code.

**Comment by flihp** [comment]
Thanks for reporting this @dilyanpalauzov. Related PRs are showing up in the thread for the IMHO high-priority items from your list.

**Comment by AndreasFuchsSIT** [comment]
bootstrap or .travis or .gitignore have no place in the dist tarball. I think this list should be stripped at places...

**Comment by diabonas** [comment]
I like the idea of trying to build from a `make dist` release tarball, since there has been at least one instance (#1243) where `make distcheck` didn't catch the distribution not containing all necessary files.

**Comment by AndreasFuchsSIT** [comment]
Sounds neat. PRs welcome.
Should maybe be a env-flag so it can run in parallel to the other tests ?

**Comment by diabonas** [comment]
> Sounds neat. PRs welcome.
> Should maybe be a env-flag so it can run in parallel to the other tests ?

Sure, I can put something together :+1: For which job(s) in the Travis matrix do we want to run this? The CI takes quite long already, so doing this for every job is probably a bit annoying time-wise.

**Comment by AndreasFuchsSIT** [comment]
We could even use this to speed up the process.
If we remove make dist from all other runs and add a new run for building+testing from tar-ball, then we should be faster in the end. (multi-thredding... :-) )

**Comment by dilyanpalauzov** [comment]
If `bootstrap` is not in the dist tarball, then it will not be possible to build the fuzzing tests from the unpacked tarball.  In turn `script/gen_fuzz.py` and `script/fuzz-log-compiler.sh` can be kept outside.

**Comment by diabonas** [comment]
> We could even use this to speed up the process.
> If we remove make dist from all other runs and add a new run for building+testing from tar-ball, then we should be faster in the end. (multi-thredding... :-) )

Interesting, `make distcheck` seems to have gotten better at detecting missing files, see the [Automake 1.15 release announcement](https://lists.gnu.org/archive/html/automake/2015-01/msg00005.html). Therefore the issue #1243 I referred to above wouldn't happen with the updated CI environment based on Ubuntu 16.04 Xenial. Hence I don't think we need to manually uncompress and build from a distribution tarball after all, since this is roughly what `make distcheck` does for us already.

I still see the appeal of moving the `make distcheck` to a separate job in order to save time on the CI. Do you want to go forward with this? The necessary changes would be https://github.com/diabonas/tpm2-tss/tree/travis-separate-distcheck-job

**Comment by diabonas** [comment]
This would roughly save a third of CI time, from about [1 hour real-time/4:15 hours total](https://travis-ci.org/tpm2-software/tpm2-tss/builds/528716148) to about [40 minutes/3 hours total](https://travis-ci.org/diabonas/tpm2-tss/builds/529049773).

Mind you, the distribution tarball might in theory depend on the configuration options (it shouldn't, of course, but you could e.g. make `EXTRA_DIST` dependent on the value of `--enable-tcti-partial-reads`). The new approach wouldn't test `make distcheck` with different configuration options.

Even if we want to keep testing these, it would probably be enough to only use `make distcheck` e.g. for the GCC jobs, saving half of the invocations.

**Comment by AndreasFuchsSIT** [comment]
Yes, you're right. We should run distcheck for all configure-options.
But restricting distcheck to gcc sounds very good.

**Comment by dilyanpalauzov** [comment]
What is considered by "make dist" should not depend on ./configure options.  The dist target considers all _SOURCEs and dist_ targets, irrespective of any AM conditionals.

There is no point to put EXTRA_DIST within AM conditional.  One way to avoid unintentionally inserting files to EXTRA_DIST within AM conditionals, is to have a single place where EXTRA_DIST is defined, without += at other places.

**Comment by AndreasFuchsSIT** [comment]
@dilyanpalauzov what is used in distcheck depends on configure variables. Thus, in order to test if anything was missing in the dist ball, we need to distcheck for different configure variables.

That can be EXTRA_DIST but more importantly header files missing in _SOURCES...

Wrt single `=` vs `+=` that's a matter of taste and I prefer the latter by far.

**Comment by diabonas** [comment]
> @dilyanpalauzov what is used in distcheck depends on configure variables. Thus, in order to test if anything was missing in the dist ball, we need to distcheck for different configure variables.

I agree it is a good idea to run `make distcheck` with different configuration options to make sure all necessary files are included. However this requires some changes to the CI since currently [only `WITH_CRYPTO` is used for `make distcheck`](https://github.com/tpm2-software/tpm2-tss/blob/9f7cc086e9802d1ed3682e8f5b036bd33dc5aa21/.travis.yml#L127), `WITH_TCTI_ASYNC` and `WITH_TCTI_PARTIAL` aren't taken into account at all.

BTW, what is the reason for having a separate `make check` and `make distcheck`? Since the tests are the thing that costs the most time, it would be desirable to get rid of the normal check and only run a distcheck. Since the latter includes the tests as well, wouldn't a `make distcheck` be enough?

**Comment by AndreasFuchsSIT** [comment]
Good question... No idea...
I guess for code-coverage collection with gcc it's better to run things in-tree instead of out-of-tree.
As for scan-build, I have no idea how that behaves when used out-of-tree in distcheck.

**Comment by diabonas** [comment]
Good point, code coverage testing doesn't work indeed with `make distcheck`. Fortunately `scan-build` does however, so new suggestion: use `distcheck` only for the Clang jobs instead of `check` and get rid of the separate `distcheck` in turn, see #1399.

**Comment by diabonas** [comment]
Note that we cannot replace `check` by `distcheck` for the [Clang fuzz target test](https://github.com/tpm2-software/tpm2-tss/blob/9f7cc086e9802d1ed3682e8f5b036bd33dc5aa21/.travis.yml#L120) because the distribution is missing some required files, which is part of what this issue is about, see https://github.com/tpm2-software/tpm2-tss/issues/1348#issuecomment-489738615:

> If `bootstrap` is not in the dist tarball, then it will not be possible to build the fuzzing tests from the unpacked tarball. In turn `script/gen_fuzz.py` and `script/fuzz-log-compiler.sh` can be kept outside.

**Comment by AndreasFuchsSIT** [comment]
IMHO, fuzz-testing is nothing to be done from the tar-ball. Thus, we can ignore it for distchecks...

**Comment by flihp** [comment]
Did a quick pass over this and it looks like the remaining files left out of the dist tarballs are as follows:
> .gitattributes
> .appveyor.yml
> doc/TSS block diagram.png
> Dockerfile
> .dockerignore
> bootstrap
> script/gen_fuzz.py
> script/fuzz-log-compiler.sh
> .travis.yml

After looking through the thread it appears as though the only remaining issue is getting `doc/TSS block diagram.png` included. My plan is to create a new PR to get this shipped and close out this ticket.
